R/pre_commit.R

```diff
@@ -0,0 +1,220 @@
+  if (length(space_files) > 0) {
+    message(
+      "ERROR: Spaces in .gitignore entries:\n",
+      paste("-", space_files, collapse = "\n")
+    )
+    all_ok <- FALSE
+  }
+
+  message("\n=== Data File Validation ===")
+  current_files <- data.frame(
+    files = list.files("./", recursive = TRUE, full.names = TRUE)
+  ) |>
+    dplyr::filter(
+      grepl(suffixes, files, ignore.case = TRUE),
+      !grepl("renv|datafiles_log.csv", files)
+    )
+
+  for (file in current_files$files) {
+    rel_path <- gsub("^./", "", file)
+
+    if (!rel_path %in% log_files$filename) {
+      message(
+        "Missing entry: ", rel_path, "\n",
+        "  - Add to datafiles_log.csv with status: ",
+        "published, reference, or dummy\n"
+      )
+      all_ok <- FALSE
+      next
+    }
+
+    status <- tolower(log_files$status[log_files$filename == rel_path])
+    in_ignore <- rel_path %in% ign_files$filename
+
+    if (!status %in% valid_statuses) {
+      if (!in_ignore && !grepl("unpublished", rel_path)) {
+        message(
+          "Unmanaged file: ", rel_path, "\n",
+          "  - Status: ", log_files$status[log_files$filename == rel_path], "\n",
+          "  - Valid statuses: ", paste(valid_statuses, collapse = ", "), "\n",
+          "  - Add to .gitignore if unpublished\n"
+        )
+        all_ok <- FALSE
+      } else {
+        message(
+          "Note: ", rel_path,
+          " is unpublished (ignored) but consider updating its status\n"
+        )
+      }
+    }
+  }
+
+  if (all_ok) {
+    message("\nAll data file checks passed!")
+    return(TRUE)
+  }
+
+  message("\nCommit blocked: Data validation issues detected")
+  return(FALSE)
+}
+
+#' Check for and Replace Google Analytics Keys
+#'
+#' Reads `google-analytics.html` and `ui.R` to check for Google Analytics keys.
+#' If found, replaces them with placeholder values and stages the files in git.
+#'
+#' @param ga_file Path to the Google Analytics file.
+#' @param ui_file Path to the UI script file.
+#' @return TRUE if no issues found, FALSE if replacements were made.
+#' @importFrom xfun gsub_file
+#' @export
+check_analytics_key <- function(ga_file = "google-analytics.html",
+                                ui_file = "ui.R") {
+  ga_pattern <- "(G-[A-Z0-9]{10}|UA-\\d{6,9}-\\d+)"
+
+  ga_content <- readLines(ga_file)
+  ui_content <- readLines(ui_file)
+
+  if (any(grepl(ga_pattern, ga_content))) {
+    xfun::gsub_file(ga_file, ga_pattern, "G-XXXXXXXXXX")
+    xfun::gsub_file(ui_file, ga_pattern, "XXXXXXXXXX")
+
+    system2("git", c("add", ga_file, ui_file))
+
+    message("Updated analytics tag")
+    return(FALSE)
+  }
+  TRUE
+}
+
+#' Check and Apply Code Styling
+#'
+#' Styles R code in the `R` directory and detects changes.
+#' If changes are made, warns the user before committing.
+#'
+#' @return TRUE if changes were made, FALSE otherwise.
+#' @importFrom digest digest
+#' @export
+style_code <- function() {
```

Comment:
This breaks if there's any sub-directories in ./R/:
- fails with `Error: The specified pathname is not a file: R/ui_panels`
- doesn't check any scripts in sub-directories
- also misses out anything in route directory - e.g. server.R, ui.R, app.R

Also doesn't feel like any of the list.files / digest code is needed. The output of `style_dir()` itself is a list of filenames and whether they've been edited by the styling, so you could just do a single line function along the lines of:

```
style_code <- function(){
   styler::style_dir()  |> 
      magrittr::extract2("changed") |> 
      any()
}
```